### ISSUE TITLE
Unexpected route behavior

### ISSUE THREAD

**Issue by monsoon235** [SagerNet/sing-box#945]
### Operating system

Linux

### System version

Ubuntu 22.04.3 LTS x86_64, kernel 6.2.0-31-generic

### Installation type

Original sing-box Command Line

### Version

<details>

Official docker image:
```console
$ docker images
REPOSITORY                                       TAG        IMAGE ID       CREATED         SIZE
ghcr.io/sagernet/sing-box                        latest     4cb3edf3fef1   3 days ago      38.8MB

$ docker run --rm -it ghcr.io/sagernet/sing-box version
sing-box version unknown

Environment: go1.21.1 linux/amd64
Tags: with_gvisor,with_quic,with_dhcp,with_wireguard,with_ech,with_utls,with_reality_server,with_clash_api,with_acme
CGO: disabled
```

</details>

### Description

Maybe it's not a bug, it's a feature. However, the description about route in the document may **lead to misunderstandings and unexpected client behavior.**

**It seems that sing-box route module will not resolve domain into IP when try to match route rules (my rules are list following).** 

Although the documentation does not explicitly state whether the domain name will be resolved to an IP in this case, it provides a **hint**, although this does not correspond to the actual behavior of sing-box.

### Reproduction

For privacy reasons, I use some domain name and IP substitutions：

- `<target_domain>`: an https server in China, with dualstack DNS response. **It allows access from China but blocks connections from oversea, and does not hit `geosite:cn`, `geosiote:geolocation-cn` etc.**
- `<target_ipv4>`: IPv4 of `<target_domain>`, **hits `geoip:cn`**
- `<target_ipv6>`: IPv6 of `<target_domain>`, **hits `geoip:cn`**
- `<proxy_server_domain>`: a shadowsocks server out of China, also with dualstack DNS response
- `<proxy_server_ipv4>`: IPv4 of `<proxy_server_domain>`
- `<proxy_server_ipv6>`: IPv6 of `<proxy_server_domain>`

I used `curl` to test:

```console
$ curl https://<target_domain>:443 --insecure  # not OK, goto proxy outbound, which is default rule
$ curl https://<target_ipv4>:443 --insecure  # OK, goto direct outbound
$ curl https://<target_ipv6>:443 --insecure  # OK, goto direct outbound
```

Client config:

```json
{
  "log": {
    "disabled": false,
    "level": "trace",
    "output": "/var/log/sing-box/box.log",
    "timestamp": true
  },
  "inbounds": [
    {
      "tag": "mixed-in",
      "type": "mixed",
      "listen": "::",
      "listen_port": 7890,
      "tcp_fast_open": true,
      "tcp_multi_path": true,
      "udp_fragment": true
    },
    {
      "tag": "socks-in",
      "type": "socks",
      "listen": "::",
      "listen_port": 10808,
      "tcp_fast_open": true,
      "tcp_multi_path": true,
      "udp_fragment": true
    },
    {
      "tag": "http-in",
      "type": "http",
      "listen": "::",
      "listen_port": 10809,
      "tcp_fast_open": true,
      "tcp_multi_path": true
    }
  ],
  "outbounds": [
    {
      "tag": "direct",
      "type": "direct",
      "tcp_multi_path": true,
      "udp_fragment": true
    },
    {
      "tag": "reject",
      "type": "block"
    },
    {
      "tag": "proxy-ss",
      "type": "shadowsocks",
      "server": "<proxy_server_domain>",
      "server_port": 114514,
      "method": "2022-blake3-aes-128-gcm",
      "password": "password",
      "tcp_multi_path": true,
      "udp_fragment": true
    },
    {
      "tag": "proxy-auto",
      "type": "selector",
      "outbounds": [
        "relay-ss"
      ]
    }
  ],
  "route": {
    "geoip": {
      "path": "/usr/local/share/sing-box/geoip.db",
      "download_url": "https://cdn.jsdelivr.net/gh/soffchen/sing-geoip@release/geoip.db",
      "download_detour": "proxy-auto"
    },
    "geosite": {
      "path": "/usr/local/share/sing-box/geosite.db",
      "download_url": "https://cdn.jsdelivr.net/gh/soffchen/sing-geosite@release/geosite.db",
      "download_detour": "proxy-auto"
    },
    "rules": [
      {  // hit oversea domains/IPs
        "geosite": [
          "gfw",
          "greatfire",
          "geolocation-!cn"
        ],
        "geoip": [
          "telegram",
          "!cn"
        ],
        "domain_suffix": [
          "google.com",
          "google.cn",
          "googleapis.com",
          "googleapis.cn",
          "xn--ngstr-lra8j.com"
        ],
        "ip_cidr": [
          "1.1.1.1/32",
          "8.8.8.8/32"
        ],
        "outbound": "proxy-auto"
      },
      {  // hit china domains/IPs
        "geosite": [
          "geolocation-cn",
          "google-cn",
          "apple-cn",
          "tld-cn",
          "cn",
          "category-games@cn"
        ],
        "geoip": [
          "cn",
          "private"
        ],
        "outbound": "direct"
      }
    ],
    "final": "proxy-auto"  // default to proxy
  },
  "dns": {
    "servers": [
      {
        "tag": "local-dns",
        "address": "local",
        "detour": "direct"
      },
      {
        "tag": "cf-dns",
        "address": "1.1.1.1",
        "detour": "proxy-auto"
      }
    ],
    "independent_cache": true,
    "rules": [
      {
        "outbound": [
          "direct"
        ],
        "server": "local-dns"
      },
      {
        "geosite": [
          "gfw",
          "greatfire",
          "geolocation-!cn"
        ],
        "domain_suffix": [
          "google.com",
          "google.cn",
          "googleapis.com",
          "googleapis.cn",
          "xn--ngstr-lra8j.com"
        ],
        "server": "cf-dns"
      },
      {
        "geosite": [
          "geolocation-cn",
          "google-cn",
          "apple-cn",
          "tld-cn",
          "cn",
          "category-games@cn"
        ],
        "server": "local-dns"
      }
    ],
    "final": "local-dns"
  },
  "experimental": {
    "clash_api": {
      "external_controller": "[::]:9090",
      "store_mode": true,
      "store_selected": true,
      "store_fakeip": true,
      "cache_file": "/var/lib/sing-box/clash_cache.db"
    }
  }
}
```

No server config because this bug does not involve the server side.

### Logs

<details>

```console
+0800 2023-10-03 00:13:32 INFO [2182137595 0ms] inbound/http[http-in]: inbound connection from 127.0.0.1:42100
+0800 2023-10-03 00:13:32 INFO [2182137595 0ms] inbound/http[http-in]: inbound connection to <target_domain>:443
+0800 2023-10-03 00:13:32 INFO [2182137595 1ms] outbound/shadowsocks[proxy-ss]: outbound connection to <target_domain>:443
+0800 2023-10-03 00:13:32 DEBUG [2182137595 1ms] dns: lookup domain <proxy_server_domain>
+0800 2023-10-03 00:13:32 DEBUG [2182137595 1ms] dns: lookup succeed for <proxy_server_domain>: <proxy_server_ipv4> <proxy_server_ipv6>
+0800 2023-10-03 00:14:03 INFO [319775608 0ms] inbound/http[http-in]: inbound connection from 127.0.0.1:47412
+0800 2023-10-03 00:14:03 INFO [319775608 1ms] inbound/http[http-in]: inbound connection to <target_ipv4>:443
+0800 2023-10-03 00:14:03 DEBUG [319775608 13ms] router: match[1] geosite=[geolocation-cn google-cn apple-cn...] geoip=[cn private] => direct
+0800 2023-10-03 00:14:03 INFO [319775608 13ms] outbound/direct[direct]: outbound connection to <target_ipv4>:443
+0800 2023-10-03 00:14:21 INFO [3879064689 0ms] inbound/http[http-in]: inbound connection from 127.0.0.1:34330
+0800 2023-10-03 00:14:21 INFO [3879064689 0ms] inbound/http[http-in]: inbound connection to <target_ipv6>:443
+0800 2023-10-03 00:14:21 DEBUG [3879064689 0ms] router: match[1] geosite=[geolocation-cn google-cn apple-cn...] geoip=[cn private] => direct
+0800 2023-10-03 00:14:21 INFO [3879064689 0ms] outbound/direct[direct]: outbound connection to <target_ipv6>:443
```

</details>

**Comment by monsoon235** [comment]
If "route module will not resolve domain into IP when try to match route rules" is indeed the designed behavior, there needs to be a reminder in the documentation.

Considering that a large number of domains are not added to the domain name list maintained by the community, I think it is still necessary to resolve the doamin in the routing module and perform matching. It can be designed similar to [domainStrategy](https://www.v2fly.org/en_US/v5/config/router.html) option in v2ray, or a better design (e.g. resolving into IPs when necessory, but holding both domain and IPs for matching).

**Comment by CHIZI-0618** [comment]
please read  https://sing-box.sagernet.org/configuration/shared/listen/#domain_strategy

**Comment by monsoon235** [comment]
Thanks for your reminder, I ignored the listening field. After I add `prefer_ipv6` in `http-in`, the domain is indeed resolved into IPv6 before routing (`www.google.com` -> `2404:6800:4005:813::2004`).

But there is a new problem: in this situation **the outboud also use IP to connect**. I mean, sing-box seems to be **unable to hold both domain and IPs, and use them in routing and dialing according to their respective `domain_strategy` settings.** If the domain is resolved to IPs, then only IPs will be available, the domain rules in the route will not work.

I think it's an inappropriate design and can be improved, ecause resolving domain need extra time, domain rules in the route can save time.

If the maintainer perfers the current design, the documentation needs to be updated to **accurately describe this behavior**. If the maintainer accepts the improvements, I may consider submitting a PR (maybe in a few months).

```
+0800 2023-10-03 17:10:22 INFO [405072927 0ms] inbound/http[http-in]: inbound connection from 127.0.0.1:36134
+0800 2023-10-03 17:10:22 INFO [405072927 1ms] inbound/http[http-in]: inbound connection to www.google.com:80
+0800 2023-10-03 17:10:22 DEBUG [405072927 1ms] dns: lookup domain www.google.com
+0800 2023-10-03 17:10:22 DEBUG [405072927 1ms] dns: match[2] domainSuffix=[google.com google.cn googleapis.com...] geosite=[gfw greatfire geolocation-!cn] => cf-dns
+0800 2023-10-03 17:10:22 INFO [405072927 1ms] dns: lookup succeed for www.google.com: 2404:6800:4005:813::2004 142.250.204.68
+0800 2023-10-03 17:10:22 DEBUG [405072927 2ms] dns: resolved [2404:6800:4005:813::2004 142.250.204.68]
+0800 2023-10-03 17:10:22 DEBUG [405072927 2ms] router: match[0] domainSuffix=[google.com google.cn googleapis.com...] geosite=[gfw greatfire geolocation-!cn] geoip=[telegram !cn] ipcidr=[1.1.1.1/32 8.8.8.8/32] => proxy-auto
+0800 2023-10-03 17:10:22 INFO [405072927 2ms] outbound/shadowsocks[proxy-ss]: outbound connection to [2404:6800:4005:813::2004]:80
```

**Comment by monsoon235** [comment]
> please read https://sing-box.sagernet.org/configuration/shared/listen/#domain_strategy

Thanks for your reminder, I ignored the listening field. After I add `prefer_ipv6` in `http-in`, the domain is indeed resolved into IPv6 before routing (`www.google.com` -> `2404:6800:4005:813::2004`).

But there is a new problem: in this situation **the outboud also use IP to connect, no matter what `domain_strategy` in dial fields is**. I mean, sing-box seems to be **unable to hold both domain and IPs, and use them in routing and dialing according to their respective `domain_strategy` settings**. If the domain is resolved to IPs, then only IPs will be available, the domain rules in the route will not work.

I think it's an inappropriate design and can be improved, because resolving domain needs extra time, domain rules in the route can save time.

If the maintainer perfers the current design, the documentation needs to be updated to **accurately describe this behavior**. If the maintainer accepts the improvements, I may consider submitting a PR (maybe in a few months).

```
+0800 2023-10-03 17:10:22 INFO [405072927 0ms] inbound/http[http-in]: inbound connection from 127.0.0.1:36134
+0800 2023-10-03 17:10:22 INFO [405072927 1ms] inbound/http[http-in]: inbound connection to www.google.com:80
+0800 2023-10-03 17:10:22 DEBUG [405072927 1ms] dns: lookup domain www.google.com
+0800 2023-10-03 17:10:22 DEBUG [405072927 1ms] dns: match[2] domainSuffix=[google.com google.cn googleapis.com...] geosite=[gfw greatfire geolocation-!cn] => cf-dns
+0800 2023-10-03 17:10:22 INFO [405072927 1ms] dns: lookup succeed for www.google.com: 2404:6800:4005:813::2004 142.250.204.68
+0800 2023-10-03 17:10:22 DEBUG [405072927 2ms] dns: resolved [2404:6800:4005:813::2004 142.250.204.68]
+0800 2023-10-03 17:10:22 DEBUG [405072927 2ms] router: match[0] domainSuffix=[google.com google.cn googleapis.com...] geosite=[gfw greatfire geolocation-!cn] geoip=[telegram !cn] ipcidr=[1.1.1.1/32 8.8.8.8/32] => proxy-auto
+0800 2023-10-03 17:10:22 INFO [405072927 2ms] outbound/shadowsocks[proxy-ss]: outbound connection to [2404:6800:4005:813::2004]:80
```
